core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -279,27 +262,31 @@ class TransactionCoordinator(brokerId: Int,
   private def logInvalidStateTransitionAndReturnError(transactionalId: String,
                                                       transactionState: TransactionState,
                                                       transactionResult: TransactionResult) = {
-    error(s"transactionalId: $transactionalId -- Current state is $transactionState, but received transaction " +
-      s"marker result: $transactionResult")
+    debug(s"TransactionalId: $transactionalId's state is $transactionState, but received transaction " +
+      s"marker result to send: $transactionResult")
     Left(Errors.INVALID_TXN_STATE)
   }
 
   def handleEndTransaction(transactionalId: String,
                            producerId: Long,
                            producerEpoch: Short,
                            txnMarkerResult: TransactionResult,
-                           responseCallback: EndTxnCallback): Unit = {
-    val error = validateTransactionalId(transactionalId)
-    if (error != Errors.NONE)
-      responseCallback(error)
+                           responseCallback: EndTxnCallback,
+                           checkTxnTimeout: Boolean = false): Unit = {
```

Comment:
Hmm, I thought it may be less efficient to traverse through the map and grab the lock on each object and check if it should be aborted, and hence I did this in two phases; plus by doing this we can also reuse the `handleEndTxn` function (note that we need to change its pending state while holding the lock), otherwise we need to do that in a separate function.

But let me know if you have a strong preference on this, then I can change it.